src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp

```diff
@@ -941,7 +941,8 @@ void ShenandoahBarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCo
     Node* gc_state    = phase->transform_later(new LoadBNode(ctrl, mem, gc_state_addr, gc_state_adr_type, TypeInt::BYTE, MemNode::unordered));
     int flags = ShenandoahHeap::HAS_FORWARDED;
     if (ShenandoahIUBarrier) {
-      flags |= ShenandoahHeap::YOUNG_MARKING;
+      assert(!ShenandoahHeap::heap()->mode()->is_generational(), "Generational mode does not support IU barrier");
```

Comment:
Not sure we need an assert here, we could check this condition in `shenandoahGenerationalMode.cpp` or `shenandoahArguments.cpp` to minimize deltas from upstream.

Reply:
Thanks.  Moving this check to shenandoahArguments.cpp.